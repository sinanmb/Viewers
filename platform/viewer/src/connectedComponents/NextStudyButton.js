import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import {
  setStudyIndex,
  selectStudy,
  setReviewStatus,
} from '../actions/workingListActions';
import { withRouter } from 'react-router-dom';
import { Icon } from '@ohif/ui';

/*
 * Next Study button marks the current study as approved before going to next study
 */
class NextStudyButton extends Component {
  handleClick = async () => {
    if (
      this.props.studyIndex + 1 <
      this.props.selectedWorkingListStudies.length
    ) {
      // Mark study as approved
      await this.props.setReviewStatus(
        this.props.selectedWorkingList,
        this.props.selectedStudy.study_instance_uid,
        true
      );

      // Go to next study
      const newIndex = this.props.studyIndex + 1;
      this.props.setStudyIndex(newIndex);
      this.props.selectStudy(this.props.selectedWorkingListStudies[newIndex]);
      const path = `/viewer/${this.props.selectedWorkingListStudies[newIndex].study_instance_uid}`;
      this.props.history.push(path);
    }
  };

  render() {
    const isLastIndex =
      this.props.studyIndex ===
      this.props.selectedWorkingListStudies.length - 1;

    const iconStyle = {
      width: '1rem',
      height: '1rem',
      color: isLastIndex ? 'gray' : 'white',
    };
    const buttonStyle = {
      background: 'transparent',
      border: 'none',
      cursor: isLastIndex ? 'default' : 'pointer',
      margin: '0.25rem',
      marginTop: '1rem',
    };

    return (
      <button
        type="button"
        style={buttonStyle}
        disabled={isLastIndex}
        onClick={this.handleClick}
      >
        <Icon name="step-forward" style={iconStyle} />
      </button>
    );
  }
}

NextStudyButton.propTypes = {
  studyIndex: PropTypes.number.isRequired,
  setStudyIndex: PropTypes.func.isRequired,
  selectedWorkingListStudies: PropTypes.array.isRequired,
  history: PropTypes.object.isRequired,
  selectStudy: PropTypes.func.isRequired,
  setReviewStatus: PropTypes.func.isRequired,
  selectedStudy: PropTypes.object.isRequired,
  selectedWorkingList: PropTypes.string.isRequired,
};

const mapStateToProps = state => ({
  studyIndex: state.workingLists.studyIndex,
  selectedWorkingListStudies: state.workingLists.selectedWorkingListStudies,
  selectedStudy: state.workingLists.selectedStudy,
  selectedWorkingList: state.workingLists.selectedWorkingList,
});

export default withRouter(
  connect(mapStateToProps, { setStudyIndex, selectStudy, setReviewStatus })(
    NextStudyButton
  )
);
