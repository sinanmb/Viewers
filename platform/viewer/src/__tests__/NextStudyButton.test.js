import React from 'react';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';

import { Router } from 'react-router';
import { MemoryRouter } from 'react-router-dom';
import { createMemoryHistory } from 'history';

import Adapter from 'enzyme-adapter-react-16';
import { configure, mount, shallow } from 'enzyme';

import NextStudyButton from '../connectedComponents/NextStudyButton';

const mockStore = configureStore([thunk]);
configure({ adapter: new Adapter() });

let store, wrapper;
const initialState = {
  workingLists: {
    selectedWorkingListStudies: [
      {
        status: false,
        study_id: 'study_id_1',
        study_instance_uid: 'study_instance_uid_1',
      },
      {
        status: false,
        study_id: 'study_id_2',
        study_instance_uid: 'study_instance_uid_2',
      },
    ],
    studyIndex: 0,
  },
};

describe('NextStudyButton component:', () => {
  beforeEach(() => {
    store = mockStore(initialState);
  });

  test('matches the snapshot', () => {
    wrapper = shallow(
      <MemoryRouter>
        <Provider store={store}>
          <NextStudyButton />
        </Provider>
      </MemoryRouter>
    );
    expect(wrapper).toMatchSnapshot();
  });

  test('it passes workingLists from the state', () => {
    wrapper = shallow(
      <MemoryRouter>
        <Provider store={store}>
          <NextStudyButton />
        </Provider>
      </MemoryRouter>
    );
    expect(wrapper.props().studyIndex).toBe(initialState.studyIndex);
    expect(wrapper.props().selectedWorkingListStudies).toBe(
      initialState.selectedWorkingListStudies
    );
  });

  test('it navigates to the next study url', async () => {
    const history = createMemoryHistory();
    wrapper = mount(
      <Router history={history}>
        <Provider store={store}>
          <NextStudyButton />
        </Provider>
      </Router>
    );

    await wrapper.find('button').simulate('click');

    const next_study_url = '/viewer/study_instance_uid_2';
    expect(history.location.pathname).toBe(next_study_url);
  });
});
