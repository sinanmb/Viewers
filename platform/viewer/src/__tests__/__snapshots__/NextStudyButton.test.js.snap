// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NextStudyButton component: Matches the snapshot 1`] = `ShallowWrapper {}`;

exports[`NextStudyButton component: matches the snapshot 1`] = `ShallowWrapper {}`;
